Correct exception specification mismatch in Memory::Recycler::AddMark

joemmett opened the pull request:
Memory::Recycler::AddMark() is declared with no exception specification
but defined as `throw()`. Upcoming versions of MSVC will diagnose this
mismatch in /permissive- or C++17 modes. Add `throw()` to the
declaration to bring the declaration in line with the definition.

Fixes #3862

Penguinwizzard commented:
Could you rebase this PR on top of master? We have fixes for the CI issues that this PR has hit in master.

joemmett commented:
Rebased, thanks!

dilijev commented:
@joemmett Sorry for the delay -- could you rebase and resolve the conflict? Then I'll get this merged.

joemmett commented:
Rebased, but it looks like the newly added AddPreciselyTracedMark() function will have exactly the same issue. I've updated the PR to change this signature as well.

dilijev commented:
@joemmett Thanks!
